src/worker/activities/activity_heartbeat_manager.rs

```diff
@@ -105,143 +99,173 @@ impl ActivityHeartbeatManager {
 }
 
 #[derive(Debug)]
-struct HbStreamState {
-    last_sent: HashMap<TaskToken, ActivityHbState>,
-    incoming_hbs: UnboundedReceiver<HBAction>,
+struct ActivityHbState {
+    last_recorded_details: Option<Vec<common::Payload>>,
+    shutdown_tx: UnboundedSender<()>,
 }
 
-impl HbStreamState {
-    fn new(incoming_hbs: UnboundedReceiver<HBAction>) -> Self {
-        Self {
-            last_sent: Default::default(),
-            incoming_hbs,
+async fn send_heartbeat(
+    sg: Arc<impl ServerGatewayApis + Send + Sync + 'static + ?Sized>,
+    cancels_tx: UnboundedSender<PendingActivityCancel>,
+    task_token: TaskToken,
+    details: Vec<common::Payload>,
+) {
+    match sg
+        .record_activity_heartbeat(task_token.clone(), details.into_payloads())
+        .await
+    {
+        Ok(RecordActivityTaskHeartbeatResponse { cancel_requested }) => {
+            if cancel_requested {
+                cancels_tx
+                    .send(PendingActivityCancel {
+                        task_token,
+                        reason: ActivityCancelReason::Cancelled,
+                    })
+                    .expect("Receive half of heartbeat cancels not blocked");
+            }
+        }
+        // Send cancels for any activity that learns its workflow already finished
+        // (which is one thing not found implies - other reasons would seem equally
+        // valid).
+        Err(s) if s.code() == tonic::Code::NotFound => {
+            cancels_tx
+                .send(PendingActivityCancel {
+                    task_token,
+                    reason: ActivityCancelReason::NotFound,
+                })
+                .expect("Receive half of heartbeat cancels not blocked");
+        }
+        Err(e) => {
+            warn!("Error when recording heartbeat: {:?}", e);
         }
     }
 }
 
-#[derive(Debug)]
-struct ActivityHbState {
-    delay: Duration,
-    last_sent: Instant,
+type HBStateMap = Arc<Mutex<HashMap<TaskToken, ActivityHbState>>>;
+type JoinHandleMap = Arc<Mutex<Option<HashMap<TaskToken, JoinHandle<()>>>>>;
+
+async fn activity_heartbeat_loop(
+    tt: TaskToken,
+    throttle_interval: Duration,
+    heartbeat_states: HBStateMap,
+    join_handles: JoinHandleMap,
+    mut shutdown_rx: UnboundedReceiver<()>,
+    sg: Arc<impl ServerGatewayApis + Send + Sync + 'static + ?Sized>,
+    cancels_tx: UnboundedSender<PendingActivityCancel>,
```

Comment:
I think this big 'ol parameter list is an example of that I mean. At minimum this could be wrapped up in a struct, but I still suspect an overall simpler take is totally doable